### ISSUE TITLE
[UX] Bring back node preview

### ISSUE THREAD

**Issue by docwilmot** [backdrop/backdrop-issues#3062]
We removed node edit previews early on, but haven't agreed on how to re-introduce it. I think this is expected functionality.
Lets decide how to bring it back.

---
~PR: https://github.com/backdrop/backdrop/pull/2174~
PR: https://github.com/backdrop/backdrop/pull/2290

**Comment by BWPanda** [comment]
I like how GitHub has a preview tab that doesn't require saving first and is easy to switch back and forth. Is something like that possible?

**Comment by findlabnet** [comment]
This is normal when content of node edited under admin theme, so we have another context of styling. What we expected to show in this preview?

**Comment by docwilmot** [comment]
@quicksketch @jenlampton was looking at the initial discussion about how D8 got live previews https://www.drupal.org/project/drupal/issues/1510544, where you both commented. We have `tempstore` in core so this is doable here. We dont have entity UUIDs though so our solution may be a bit more volatile. I was considering:
- on node preview assign a tempstore ID based on `uid` and node type
- if the user navigates away then when he attempts to create a new node of that type load from tempstore and require either save or cancel before allowing a new node.

Would that work?

**Comment by docwilmot** [comment]
Did it, mostly copying Drupal.
Still very preliminary but nothing complex left to do I'd think.
The cancel button is broken with this.

**Comment by docwilmot** [comment]
@jenlampton here's a (cheeky) suggestion. The issue of poor previews in D8 was considered a critical bug. Can we consider no previews at all in Backdrop also a critical bug and fix it for 1.9?
(This PR isnt ready; I only pushed it because I seemed to have messed up my local repo somehow and it was past my bedtime. But the rest of the work is mostly cosmetic, and can finish this weekend.)

**Comment by jenlampton** [comment]
My initial feedback is very positive!  The preview works as expected, and in the correct front-end theme! It needs some UX cleanup, but it's a lot closer than I expected :) Nice work @docwilmot!!

I'm not sure how much my initial reaction has been colored by past drupal experience (probably a lot) but here's my initial feedback:

1) The positioning of the preview button worried me initially. I'm not sure why - that's where it used to be...  but I was looking for it at the top of the page for some reason? Not sure if this is valid feedback but leaving it here incase others experience it too. 

2) when I clicked the preview button, I was surprised that I left the page I was editing, and I was worried that all my changes had been lost, or were about to be lost if I did something wrong. We may want to add a message like we do for views & layouts letting people know that their changes have been saved temporarily. 

2a) I was initially expecting either a modal preview, or something to appear on the same page, but after playing with this a little, I actually like it better, as what I'm looking at is *almost* what it will look like when it's published.

2b) In order to make it clearer what on the page is real content, and what on the page is editorial tooling, we should style the .node-preview-container differently, somehow. 

3) I wanted a "save" button directly on the preview page. If I'm happy with what I see I don't want to go back to the node form in order to save the node.

4) The language & location of the return to editing link was unclear to me. Maybe if it looked like a button that would be better?

5) the display-modes drop-down needs a label

6) we should remove the search and RSS display modes. This is a larger/separate problem See https://github.com/backdrop/backdrop-issues/issues/3082

7) This is the doozy: What about layouts? If a node is to be rendered in a 3 column layout with a field in each column, is that taken into account in the preview? It needs to be, but I think that makes the display-mode switcher approach less viable...

**Comment by docwilmot** [comment]
Didnt think you'd fall for that. 
The doozy is something I didnt consider. Good point. Should be doable I think, will have a look.
The other significant point is the position of the `.node-preview-container`. It is in fact supposed to be at the top of the page. In D8 they have `hook_page_build` to position stuff on the page, but we dont. I was evaluating hacking Standard renderer to place it. I'll wait for other ideas from @quicksketch  and anyone else.
Other things you list are/were in the plan, but since we're taking 4 more months to get this done, no rush.
> I was initially expecting either a modal preview, or something to appear on the same page

Ughh, that's like so 2006! 😄

**Comment by olafgrabienski** [comment]
One small observation: I was previewing a post (let's say "Post One") and noticed its path `node/post/preview` (without a node ID). For curiosity I tried to add another post without canceling or saving the previewed post before. Instead of adding a new post I ended up in the form of "Post one", including the message `Continue editing this post or cancel to clear.`. Can we assign a (temporary) node ID for the preview to avoid this conflict?

**Comment by docwilmot** [comment]
@olafgrabienski good point, and this was considered, as mentioned in https://github.com/backdrop/backdrop-issues/issues/3062#issuecomment-386512432 above. For a new node, without yet a node ID, we could assign a random unique ID (the initial D8 patch just used `time()`). 

But then if you navigated away from the preview, unless you remembered what the unique ID was (unlikely if is `9842946248` or `BY87-87f7-f65f4-FRT8`), you'd lose your work. So we'd need to add an administrative list of incomplete nodes, and/or a garbage collection to destroy old half-finished nodes on cron. We'd be treating these like revisions rather than just a preview, and this would be more complicated.

This would be a reasonable alternative, as it would make 'previews' more disposable - if you dont like what you see, just `node/add` and start again. But it is also more likely to cause alarm when someone thinks theyve lost their work if the unintentionally navigate away.

This PR probably doesnt help much with the initial alarm but at least a new user would quickly learn, if I navigate away, just go back to `node/add` and continue.

Advantages both ways, but less complexity this way I feel. Other opinions welcome though. @klonos?

**Comment by olafgrabienski** [comment]
@docwilmot I see the advantage of the current PR and the hurdles regarding the node ID.

Actually, I was mainly concerned about conflicts between different editors: What would happen if Editor A adds "Page one" and previews it while in the meantime Editor B adds another page. I guessed that Editor B would also end up in the "Page one" form of Editor A. But I've tested it now in the sandbox site, and everything seems fine: Editor B is able to add a page in its own form. (How does that work, btw?)

**Comment by docwilmot** [comment]
> Actually, I was mainly concerned about conflicts between different editors

Understood.
The PR adds an entry in tempstore with an ID based on the UID and the Node type. So every user can have one in-preview node at any time.

**Comment by docwilmot** [comment]
>This is the doozy: What about layouts? If a node is to be rendered in a 3 column layout with a field in each column, is that taken into account in the preview?

@jenlampton @quicksketch the only way (I can think of) to get a path to use a layout that is not explicitly defined for it (ie programmatically force a different layout) is, coincidentally, to use the new `$renderer_name` property we've added to the Layout class; for this example:
- manually create a layout config file for `node/%/preview` in the config active directory
- add a custom `renderer_name` and create the custom Renderer
- in that new renderer,  override `renderLayout()` to check if the layout you want to use exists (check if a layout for `node/%` exists in this case) and use that layout instead.

This should work normally but I'm not sure if our preview URLs of `node/NODE_ID/preview` would actually resolve to `node/%/preview` int this case. But this isn't the point of this shout out to you.

I am actually wondering if special-casing node previews in Layout module would be simpler than all this, plus it really wouldnt do to have the preview layout in the layout list anyway.

Or if you had other ideas.

**Comment by quicksketch** [comment]
I haven't tried out the sandbox yet, it sounds rather exciting though. :)

Could we simply assign the node an ID when the preview is initiated? In WordPress, drafts are saved automatically as the content is created. I would think that a combination of auto-saving and then just showing the unpublished node page would achieve a nearly perfect preview. Essentially just streamline the normal approach of posting an unpublished node, viewing it, and then publishing it.

We currently have some problems in Backdrop that required fields *must* be entered before the form can be saved. Ideally we'd find a way to correct the dependence on the validation layer to tidy-up the data format, and make it so that content could auto-saved (or previewed) even before it was complete. Full validation of data on input would only be required when publishing the content.

**Comment by laryn** [comment]
@quicksketch FWIW the Save Draft module has an option to allow drafts to be saved even if required fields are not entered. Not sure if that is helpful here or not.

**Comment by docwilmot** [comment]
> auto-saving

Oh wow, didn't think you'd lean that way after reading that long issue on Drupal. This would essentially be Save Draft in core as @laryn suspects, but with a "Preview" button instead.

Not sure the implications.

**Comment by olafgrabienski** [comment]
> I would think that a combination of auto-saving and then just showing the unpublished node page would achieve a nearly perfect preview.

Personally, I often just visit the unpublished node page instead of using a preview, so I like this approach very much.

**Comment by docwilmot** [comment]
Had a look at how Wordpress does things. They have revisions in multiple states and the autosave saves in a "preview" state, not as live nodes. We dont have that, and we have no way to let modules know that our "preview" is not a save.

I personally dont like the idea of saving nodes before they are ready. Thinking of all the contrib modules which are set to do things when a node is created: messaging modules, ecommerce modules, Rules, and goodness knows what else all firing off when you prview a node. To start having users "saving" nodes without knowing that this is what theyre actually doing seems dangerous to me. This is something which should either follow an update to our revisions system (see https://www.drupal.org/project/drupal/issues/218755?) or in v2.0 when we can warn module builders of such a significant change.

My PR begins to provide essentially the same ability to view a node in close to the live state, and we can do it now.

**Comment by jenlampton** [comment]
> My PR begins to provide essentially the same ability to view a node in close to the live state

It was the "close to" problem that got previews removed from core in the first place. We've got to figure out the layout bit in one way or other before this can get in. I'm also on the fence about adding auto-save before 2.0. Maybe we can fake it in another way to generate a node context for layouts to use?

**Comment by docwilmot** [comment]
Note I'm not even defending this PR or suggesting its the right way. I'm open to suggestions. But using the Tempstore approach is what D8 did, and it works well; it is essentially the same node viewed in the required view mode. Thats no where close to the madness of D7 previews.

> Maybe we can fake it in another way to generate a node context for layouts to use?

This is actually what I was asking above about special-casing. We could circumvent all of the custom `renderer_name` shenanigans I mentioned by just tweaking the layout routing system itself, example: 
```
function layout_get_layout_by_path($path = NULL, $router_item = NULL) {
  ...
  if (empty($cache[$router_item['path']])) {
    // If the path is a node preview path, check if a custom layout exists for
    // nodes and use that instead.
    if ($router_item['path'] == 'node/%/preview') { 
      $layouts = layout_load_multiple_by_path('node/%', TRUE);
    }
    else {
      $layouts = layout_load_multiple_by_path($router_item['path'], TRUE);
      ...

```
That would work.

**Comment by docwilmot** [comment]
Fixed so preview now uses a node layout if available.

**Comment by docwilmot** [comment]
To allow anon to do previews, might be best to store the preview ID in `$_SESSION`, which makes me think should use that for everybody, instead of linking the ID to node type and UID. Thinking.

**Comment by docwilmot** [comment]
Mentioned in meeting today, so just thought I'd chime in: the reason why this is paused isn't because we're trying to make it work with layouts, because I fixed that weeks ago, but to have confirmation as to if we're going with the current `Tempstore` approach, or the partial save approach @quicksketch suggested in comment https://github.com/backdrop/backdrop-issues/issues/3062#issuecomment-387276470. I'd please encourage to check the latest sandbox as well.

**Comment by jenlampton** [comment]
I'd vote for the tempstore approach, but what do others think?

**Comment by herbdool** [comment]
Tempstore sounds like best, most BC approach.

Regarding anonymous users and tempstore, D8 has an issue about that: https://www.drupal.org/project/drupal/issues/2743931

**Comment by docwilmot** [comment]
- added label for view mode button 
- added a save button on the preview bar.

**Comment by Graham-72** [comment]
I have briefly tried the sandbox and I like this. 
A couple of comments:
The 'back to content editing' link perhaps needs a bit of styling, though it certainly stands out as it is, which is good.
I ended with two advisory notices, one of which was 'Continue editing this post or cancel to clear' which was no longer relevant. I think that to avoid confusion only the currently applicable notices should appear. I have noticed this happening in other BD situations, so is this a natural feature of BD?

**Comment by olafgrabienski** [comment]
Nice progress! Is it possible that the view mode dropdown displays the chosen value? At the moment it displays always "Default" even if another view mode has been chosen.

> The 'back to content editing' link perhaps needs a bit of styling ...

I'd suggest to style the back-to-content-link more simple, to remove the 'arrows' and to adjust the alignment of some elements - see the following screenshots:

**Current:**

![screen-backdrop-pr-2174](https://user-images.githubusercontent.com/14188307/41820932-afac0194-77d9-11e8-85ff-bafae19577ae.png)

**Proposal:**

![screen-backdrop-pr-2174-additional-styling](https://user-images.githubusercontent.com/14188307/41820940-bf107778-77d9-11e8-8d0e-00faa73d097f.png)

**Comment by docwilmot** [comment]
PR updated.
One thing the D8 proposal for this had was a JS technique to remove all links on the page so that the user would not inadvertently migrate from the preview. This was important in D8 because they used the node UUID to build the preview URL, like `node/<UUID>/preview` which is of course impossible to remember, so if you clicked a link you might lose your data. This PR doesnt do that, so you can navigate away, but once you retrun to `node/add/<NODE TYPE` your data will be restored from tempstore. 

But we could still kill all links, to at least prevent scaring our users. Thoughts?

**Comment by jenlampton** [comment]
> But we could still kill all links, to at least prevent scaring our users.

I think this is a good idea :)

**Comment by klonos** [comment]
Great work @docwilmot!! ...as always 😄 

It seems that I am coming late to the party, so sorry if I am reiterating comments made by others earlier. I have read all your comments guys, but it seems that I have missed some earlier versions of the PR sandbox, so I might be missing something at times.

Here's my thoughts after playing around with the latest iteration:

1. This introduces a new ribbon/bar that has controls for saving/cancelling something temporary that has not been saved yet. We are using this same UI pattern in various places in the backend (views being the more obvious thing that comes to mind), so we should aim for some consistency. In the spirit of this consistency, I would expect a message to be present here: "* This is a preview, and all changes are stored temporarily. Click Save to make your changes permanent. Click Cancel to discard your changes."

2. The before-mentioned ribbon is being rendered at the top of the page, and by default is not sticky. So when the user scrolls down, it gets out of their viewport. My workflow while previewing would be to scroll from the top of the page, to the bottom of the page. So naturally, I would expect that when I reach the bottom of the page, controls will be made available to me to finish the task (either save, or cancel and return back to edit mode). In other words, I would expect this preview bar to be at the bottom. Now, if we do place it at the bottom, we need to make it sticky too I believe.

3. Styling this ribbon to fit the frontend theme makes it look as if it is part of it, whereas I would expect to have some separation here. Once again, I will shamelessly plug @laryn's solution from https://github.com/backdrop/backdrop-issues/issues/566#issuecomment-384853607 :

    ![screen shot 2018-04-26 at 10 25 10 pm](https://user-images.githubusercontent.com/871421/39343135-e9350008-49a0-11e8-8b2f-439c0cc98ce7.jpg)

    ...having the rest of the page be styled with the frontend theme, while this bit sits at the bottom and is rendered using the admin theme would clearly convey which bits are part of the actual page the anonymous visitors get to see, and which is part of the admin controls that save or discard the preview currently being displayed.

4. Can we please make the "Select a display mode" dropdown autosubmit, and remove the "Switch" button? It's just a single action (as opposed to the exposed filters in the content management page for example), so no need for extra clicks when we can avoid them.

5. Making the admin bar sticky at the top, retains the top margin of the page, so the "Back to content editing" link in the new ribbon does not get hidden behind the admin bar 👍 ...not part of the changes in this PR, nevertheless good to find that this has been future-proofed, and I thought I'd mention that this has been tested 😄 

6. Trying to create another piece of content of the same type always returns you to the preview of the last piece of content you were previewing. I understand that this has been a technical decision made early on (@docwilmot, I have read your comment re lack of UUID etc.), but as a person that is constantly multitasking and has multiple browser tabs open at each given time, I find this very limiting. I need the following:

a) be able to start drafting multiple pieces of content at the same time.
b) have a way to return to half-finished drafts of pieces of content (where I have clicked "Preview", but have left the tab without saving nor discarding changes).

7. "Continue editing this page or cancel to clear." should perhaps be "Recent changes you have made to this [content_type] have not been saved. You may continue editing and save changes, or click Cancel to discard them." instead.

8. Is the "Save" button on the preview page a different thing than the submit buttons in the edit form? If so, then this should not be the case, because:

    a) we should be reusing code anyway as best practice (I am assuming the we can just hide the "Preview" button if the current page is a preview, and hide the "Back to content editing" link when this is NOT a preview page)
    b) it will make things easier when more submit buttons are added either by contrib or by core in the future

9. The  "Back to content editing" link is the respective "cancel" link. It should be to the right of the "Save" button; not on top of it.

10. Bug: Publishing option changes to "Schedule for later" when returning back to edit mode from preview.

11. **Data loss** when opening another tab to create same type of content. STR:

    - start creating a page
    - add a page title and some body text
    - without changing anything else hit the "Preview" button.
    - while keeping the preview open in the current browser tab, try creating another page on another browser tab -> you are taken to a page that has the title and the body text of the previous page you were creating (👍), but the publishing options have defaulted to "Schedule for later" (👎)
    - close the second tab **without making any changes**
    - back to the first tab with the preview, click on the "Back to content editing" link -> publishing setting is restored to "Publish now" (👍), but the page title and the body text are now gone (👎)

    ...close all tabs and repeat the above steps, but this time when creating the page switch to draft before hitting "Preview" ..... -> publishing setting is restored to "Publish now" (this time this is a 👎), and the page title and body are also gone (👎)

    I have not tried changing other settings or adding things to other types of fields, but I am assuming that in this scenario data/configuration is just not retained.

That's it for now 😄

**Comment by docwilmot** [comment]
@klonos 1-5 good ideas, will do.

6 has been mentioned by @herbdool https://github.com/backdrop/backdrop-issues/issues/3062#issuecomment-391758462. Two people mentioning makes me suspect it probably isnt too unusual a practice then. If we allow multiple items per user in tempstore we will need some way to make each preview unique by assigning a unique ID in the URL (eg `node/5/preview/6df65f16d5fv9df`). This means if you accidentally close the tab or window, you'll lose your preview unless you remember the ID (`6df65f16d5fv9df`), which isn't any worse than what we have now (probably better, actually; contrib could probably even come up with a Lost Preview Module). So OK, then, will work on that.

7 OK; will be less necessary if I fix 6 above, since when you return to a node edit page, you wont have half finished edits popping up, which is why this message was necessary to start with.

8-9 No its not the same button , and yes it does re-use the node save code (it runs `node_form_submit()`). But if 2-3 in your list are accomplished, agreed it will need to look like the node edit submit buttons.

10 Good catch

11 Wont matter if point 6 is fixed.

Thanks for the review.

**Comment by klonos** [comment]
Thanks for the hard work @docwilmot 👍...looking forward to test the next iteration.

re 6: yeah, as explained, I am aware that this requires certain things like UUID (or any form of id we implement), but yes, not that uncommon 😄

re 7 and 11: fair enough. I made the comments and did the testing because I was not sure that 6 would be accepted easily 😄

re 8-9 👍

**Comment by docwilmot** [comment]
Updated, except the "schedule for later" bug, which will wait till after daddy duties.

**Comment by docwilmot** [comment]
![preview](https://user-images.githubusercontent.com/1042283/42127998-182e8a2e-7c9a-11e8-9a21-c953ee1ce439.PNG)

**Comment by klonos** [comment]
Looks awesome @docwilmot! 👍So, comments:

1. This looks more in par with the rest of the I now 👍. We might need to tweak the wording of the message a bit to add something along the lines of "no changes are visible to site visitors yet". This might sound superfluous given the "changes are stored temporarily" bit, but it would help clarify things, because people that are used to the Drupal/Backdrop ways will suddenly see something being rendered in the frontend theme, and they might get the wrong idea that their changes have gone public. Perhaps after #3155 is merged we can make this an `info` type of message, to differentiate a bit more from the green "Things have been saved" `status` message.

2 & 3. This looks brilliant! 👍. A few comments/thoughts...

    a) From the PR it seems that the sticky ribbon with the form submit buttons at the bottom lives in `node` module (its styling at least). I think that it should be either in `system`, or `form` modules instead(?), so that other modules like views etc. can benefit from it. Point is that this should not be specific to this feature here; it should be something system-wide.

    b) Once you have scrolled to the bottom of the preview page, you can see the tip of Drop's head, meaning that there is a portion of the page that is hidden:

    <img width="351" alt="screen shot 2018-07-03 at 9 09 45 am" src="https://user-images.githubusercontent.com/2423362/42190644-9a5d0120-7ea1-11e8-99ae-09436be7a6fa.png">

    ...we should perhaps use the same trick that we have in place for the admin bar.

    c) The submit buttons and the display mode switcher do not look good floated left/right on wider screens:

    <img width="1666" alt="screen shot 2018-07-03 at 9 10 33 am" src="https://user-images.githubusercontent.com/2423362/42190899-31858e72-7ea3-11e8-9247-6e381c474939.png">

    ...they should be "contained" to the same width of the `.l-wrapper-inner container` div and some spacing added to them (also perhaps inline the label of the switcher with the drop-down). Something like this perhaps:

    ![screen shot 2018-07-03 at 9 24 11 am](https://user-images.githubusercontent.com/2423362/42190941-65f23a5c-7ea3-11e8-8a23-f14152c869c0.png)

4. Poifect! 👍 

5. N/A any more.

6. Also 👍 ...we can either consider offering a view that shows half-done content (in cases where browser crashes and tabs are lost), or leave that for contrib.

7. Hitting "back to edit" and then also "Preview" a second time retains the UUID 👍 

8. I installed the [Save Draft](https://github.com/backdrop-contrib/save_draft) contrib module in the PR sandbox. I expected the "Save as draft" button to be available during preview. It was not 👎 

9. The "Back to content editing" link sits to the right of the "Save" button instead of on top of it. Yay consistency! 👍 

10 & 11. No options reset/change bug - no data loss 👍

**Comment by klonos** [comment]
...re disallowing clicks when in preview mode, the admin bar links keep working (understandable?), while contextual links do not work (should they?). In general, should we disable the links, or perhaps make them all open in a new tab?

**Comment by klonos** [comment]
...on mobile, the submit ribbon (we need to find a name for this 😄 ) could use some styling:

![screen shot 2018-07-03 at 9 51 27 am](https://user-images.githubusercontent.com/2423362/42191548-cdf1f9e6-7ea6-11e8-81f1-b502be919d23.png)

**Comment by docwilmot** [comment]
re 2-3,  a bit of a disclaimer about this ribbon: it is using fake Seven styling for now. I cant force the ribbon to use the admin theme styles on a non-admin page. Here's why: realistically, any theme can be an admin theme so would need to include the CSS for `.node-preview-container`. So if  the ribbon loads the admin theme, then when the preview page loads, the CSS from Basis would load, **and** the CSS from Seven would have to load too. You'd have two `.node-preview-container` declarations, and no way to decide which would win. 

Except maybe if in every theme's `template.php` we have code that checks the path, and loads that theme's `node-preview.css` only if it is being used as admin. Which seems like a lot of work for not much gain.

Also, before this PR I had the ribbon in blue (it wasn't using Basis as you thought, just coincidentally a random blue that looked similar). I am in favor of having the ribbon actually be a more dramatic color to let the user know that this isnt a normal part of the page.
> I think that it should be either in system, or form modules instead

This ribbon is a very simple `DIV` popped on the page with jquery; it isnt a standard form element or anything, so no, it shouldnt be in `system` module. When #566 drops we could probably use that code, but not sure. 

> you can see the tip of Drop's head

This works fine on my local computer, dont know why the sandbox wont cooperate. Might need some CSS guru help with that.

> they should be "contained" to the same width of the .l-wrapper-inner container div

Can add an inner container to the ribbon but see point above; this only helps if the front end theme is using a `.container` class, as Basis does.

> I expected the "Save as draft" button to be available during preview

No, the ribbon has nothing to do with FormAPI actions; thats #566, different issue.

> In general, should we disable the links, or perhaps make them all open in a new tab?

Group decision I'd say. But I think disable.

> on mobile, the submit ribbon could use some styling

Yep

**Comment by klonos** [comment]
Thanks for the explanation re the ribbon theme @docwilmot 👍 ...it makes much more sense now. I don't like the fact that the backend submit buttons are different than what we have on this preview ribbon, but it is what it is. Agreed that we should revisit this once #566 lands.

All the rest are perfectly valid points too. I guess there are some limitations that hinder making this work ideally. We'll just have to settle for less for now. Lets improve iteratively 😄

**Comment by klonos** [comment]
I believe that the general consensus during the last dev meeting is that we should not be unnecessarily  holding this issue up.

...if contrib modules like save_draft want to add their buttons in the node preview ribbon, then they should take care of doing that. I personally would like to get this done too because a) we should be favouring contrib over core (philosophy matter), and b) it seems that we will most likely will include the save_draft functionality in core, so if we do not deal with this now, we'll have to deal with it later as a core UX problem.

>I'm actually in favor now of styling this ribbon uniquely so that its obvious that it isnt Seven, to ease any confusion. This was the plan in the D7/8 version, and I think it makes sense.

IMO, the technically correct direction here would be to have the ribbon styled with whatever theme is set a default admin theme in `/admin/appearance` (`"admin_theme"` value in `system.core.json`); not having it styled as a specific theme, because that would look weird if for example default admin theme was Bartik, default FE theme was Basis, and the ribbon was styled as Seven. Just saying.

In any way, lets just improve iteratively and get this awesome feature in 1.11 😄. What else is left to be done here?

**Comment by herbdool** [comment]
@docwilmot I found a bug:

1. Go to `/admin/content` and click `Edit` on a node.
2. On the edit screen click `Preview`.
3. Sends me back to the content overview:

![screenshot from 2018-08-04 08-36-40](https://user-images.githubusercontent.com/1071483/43676553-93418e36-97c1-11e8-82eb-df976f439f82.png)

**Comment by herbdool** [comment]
@klonos what do you think of my idea: a cleaner way of not holding up this issue is to get rid of the `Save` button the preview ribbon altogether. Maybe the only thing we need is the `Back to content editing` link (and the `Select a display mode` dropdown). I like this approach since the danger of having contrib modules putting their action buttons and settings on the ribbon will be cramming too many options into the ribbon and just becomes a duplicate of the edit screen.

We can always revisit the ribbon to add more buttons, but it's harder to take them away once we've got them there.

**Comment by klonos** [comment]
I see the reasoning behind your suggestion @herbdool, but the point is to provide a good UX, and adding clicks to something that can be done with less click is not helping achieve good UX. I realise that this issue is being help up, but I also have in mind what @mikemccaffrey said at some point during a dev meeting: "we should not ship broken things with core". This is not broken per se, but if we were to remove the save button, it would be incomplete compared to the D7 preview functionality.

I would like to hear what others think, but I would prefer to have the save button, and if/when #566 lands, we can revise the code. From an end user PoV, it would not matter what we have done code-wise to achieve the save button (or any other button) being available in the ribbon; it would simply be a cool feature they could use, with the be UX possible at the time.

**Comment by herbdool** [comment]
I lean more towards "Release early, release often" https://en.m.wikipedia.org/wiki/Release_early,_release_often. If it's not broken, which this isn't, and it meets a minimal viable solution, then let's release it and get real world feedback.

**Comment by klonos** [comment]
Absolutely! ...we do have time till code freeze though 😄

**Comment by olafgrabienski** [comment]
> What else is left to be done here?

As a reminder: remove the search and RSS display modes, see #3082.

**Comment by olafgrabienski** [comment]
> I found a bug: Go to /admin/content and click Edit on a node. On the edit screen click Preview. Sends me back to the content overview

The same happens with contextual links: go to the home page, use a contextual "Edit" link to edit a post, click "Preview". Sent back to the home page, I see the "This is a preview..." message but no ribbon. I guess that's because of the `edit?destination=home` part of the "Edit" link.

**Comment by docwilmot** [comment]
This is ready for review now. Tests added and all concerns addressed. A date related fail isn't happening on my local machine and I suspect it's not related. Thanks for all suggestions everyone.

**Comment by herbdool** [comment]
The destination issues are fixed. Links in the node are disabled. What about the rest of links? I'm of two minds. Maybe they need to stay clickable but it seems a bit odd when it is a preview.

**Comment by herbdool** [comment]
I tested having two node previews at the same time. That looks okay too.

**Comment by herbdool** [comment]
The ribbon isn't forgiving on smaller screens. The buttons get pushed off the screen.
![screenshot_20180806-135025 01](https://user-images.githubusercontent.com/1071483/43732615-29a3a3a4-9980-11e8-9aa8-243fcaf1d350.jpg)

**Comment by docwilmot** [comment]
@herbdool thanks. will need a design review I think. not my strong point to be honest. maybe can get @wesruv to have a look?

**Comment by herbdool** [comment]
Might just need a couple media queries that match the breakpoints in core. And increase the height if on smaller screens.

**Comment by olafgrabienski** [comment]
@herbdool Simple and good approach. There's an issue though on very small screens like 320px (e.g. iPhone 4, 5 and SE) where the width isn't sufficient to display the "Back to content editing" in one line (see screenshots below). To fix that, we could remove `margin: 10px` from `.node-preview-backlink`, or set `margin-right` to `0`.

See in comparison:

**360px (ok):**

![screen-backdrop-pr-2174-360px](https://user-images.githubusercontent.com/14188307/44024560-e9d1a4b0-9eed-11e8-8fc8-88bd43385335.png)

**320px (looks broken):**

![screen-backdrop-pr-2174-320px](https://user-images.githubusercontent.com/14188307/44024581-f88d202e-9eed-11e8-9237-f0bc78f1e197.png)

**Comment by docwilmot** [comment]
Thanks for the reviews. Would also be grateful for your opinions on whether we should keep the Seven-style theming for the ribbon, or give it its own unique colors and style. My concern is that I would want to be sure that its obvious that a user is not on just a regular node page and they dont miss the ribbon with such subtle styling as it has now. Maybe a blue background, or grey etc?

Also, note the discussion above: its fake Seven, and will persist even if you change your admin theme to something else; we need to make sure the style is generic enough.

**Comment by olafgrabienski** [comment]
> To fix that, we could remove margin: 10px from .node-preview-backlink, or set margin-right to 0.

Or better, on small screens, let's display the 'left' and the 'right' part of the form one below the other:

![screen-backdrop-pr-2174-320px-left-right-below-each-other](https://user-images.githubusercontent.com/14188307/44027098-8b1f780e-9ef5-11e8-9a21-eace61c45644.png)

**CSS for small screens:**

```
div.node-node-preview-banner-form-left {
  /* float: left; */
  display: block;
}
.node-preview-backlink {
  margin: 10px 0 10px 10px;
}
div.node-node-preview-banner-form-right {
  /* float: right; */
  display: block;
}
```

(Need the `div` part to overwrite `.container-inline div { display: inline; }`.)

**Comment by olafgrabienski** [comment]
> Would also be grateful for your opinions on whether we should keep the Seven-style theming for the ribbon, or give it its own unique colors and style. My concern is that I would want to be sure that its obvious that a user is not on just a regular node page and they dont miss the ribbon with such subtle styling as it has now. Maybe a blue background, or grey etc?

In my opinion the blue background -- visible in the screenshot of https://github.com/backdrop/backdrop-issues/issues/3062#issuecomment-399767670 -- is a good idea.

**Comment by klonos** [comment]
I on the other hand think that eventually we should be using a technique similar to what we are implementing with the theme color settings preview (#1949), and that the ribbon should be true backend/admin theme (also in order to possibly make it easier to accomodate for additional submit buttons added by contrib). But all these can be follow-ups.

**Comment by herbdool** [comment]
I've updated my snippet based on @olafgrabienski suggestions:
```
.node-preview-container {
  font: normal 16px/1.538 "Open Sans", "Lucida Grande", "Lucida Sans Unicode", sans-serif;
  background-image: -webkit-linear-gradient(top, #d1e8f5, #d3e8f4);
  background-image: -moz-linear-gradient(top, #d1e8f5, #d3e8f4);
  background-image: -o-linear-gradient(top, #d1e8f5, #d3e8f4);
  background-image: linear-gradient(to bottom, #d1e8f5, #d3e8f4);
  box-shadow: 0px 2px 20px rgba(0,0,0,0.4);
  -webkit-box-shadow: 0 1px 3px 1px rgba(0, 0, 0, 0.3333);
  z-index: 999;
  width: 100%;
  padding: 10px;
  overflow: auto;
  position: fixed;
  bottom: 0;
  height: 160px;
}
@media (min-width: 34em) {
  .node-preview-container {
    height: 120px;
  }
}
.page {
  padding-bottom: 160px;
}
@media (min-width: 34em) {
  .page {
    padding-bottom: 120px;
  }
}
.node-preview-backlink {
  text-decoration: none;
  margin: 10px 0 10px 10px;
}
@media (min-width: 34em) {
  .node-preview-backlink {
    margin: 10px;
  }
}
div.node-node-preview-banner-form-left {
  display: block;
}
@media (min-width: 34em) {
  .node-node-preview-banner-form-left {
    float: left;
  }
}
.node-node-preview-banner-form-right {
  display: block;
}
@media (min-width: 34em) {
  div.node-node-preview-banner-form-right {
    float: right;
  }
}
```
I've only tested using the Chrome inspector.

We should make sure the blue background and text meets accessibility guidelines.

**Comment by herbdool** [comment]
Just checked with https://webaim.org/resources/contrastchecker/ and sure enough it fails. We should probably set the link colour as well if it's all going to be unique from Basis.

**Comment by herbdool** [comment]
One more thing, I recommend renaming `node-node-preview-banner-form-right` and `node-node-preview-banner-form-left` since aren't necessarily going to be on the right and left.

**Comment by docwilmot** [comment]
New commit addresses above issues.
- renamed themed ribbon sections
- recolored to a neutral grey color that is neither Seven nor Basis so there is no confusion about the origin of this ribbon. Its a simple DIV placed there with JS, thats it,
- update styling to accommodate for no-JS situations.

**Comment by docwilmot** [comment]
Please RTBC if you review and are happy.

**Comment by herbdool** [comment]
@docwilmot it's looking good. Mobile looks better. I left a code review.

**Comment by docwilmot** [comment]
Checked your comments and would be happy for another review.

**Comment by olafgrabienski** [comment]
I've had a look at the sandbox site, only regarding the look and feel which is. good in my opinion, also on mobile. I guess you removed some styles (e.g. the box shadow) to keep it simple which looks indeed better.

Only suggestion (definitely not a blocker): add a padding-top of 6px to `node-preview-banner-form-actions` for not sticking the element too close at the outer, and for better alignment with the other element.

**Comment by docwilmot** [comment]
@olafgrabienski can you mark this as RTBC then please, so @quicksketch can give it a final lookover?

**Comment by olafgrabienski** [comment]
Sure, done!

**Comment by herbdool** [comment]
@docwilmot I made some code review comments. All small fixes. Looking close.

**Comment by docwilmot** [comment]
@herbdool all done

**Comment by herbdool** [comment]
Awesome!

**Comment by herbdool** [comment]
@docwilmot I just discovered an issue.

Steps to reproduce:
1. Set page content type to be "draft" by default.
2. As a non-admin user add a page
3. Before saving click preview.
4. Then save.

You will see that the node is published when it should be unpublished.

This doesn't happen if you skip preview and just save it.

**Comment by herbdool** [comment]
Good.

By the way, I made a couple small comments. One, because I had just been working on the "pageless" PR and wondered how node preview would handle it.

**Comment by docwilmot** [comment]
Thanks for the comments; all updated again.

**Comment by herbdool** [comment]
@docwilmot better but still an issue:

1. Add page and select scheduled.
2. Click preview and save from preview screen.
3. Edit to view settings. The page will be set to draft rather than scheduled.

**Comment by docwilmot** [comment]
@herbdool has been very good at breaking this. Next round sir. Or @quicksketch if you're ready to have a look.

**Comment by quicksketch** [comment]
Overall, PR looks good. I posted a code review at https://github.com/backdrop/backdrop/pull/2174#pullrequestreview-151578246

Pulling out the main summary from there:

> My only big hangup is the strong cross-connection between Layout and Node that is added here in `layout_get_layout_by_path()`. Let's make sure we outline decoupling these in an issue and include a `@todo`.
>
>There are still a few tests failures that need to be addressed as well.
>
> Let's also add `@since 1.11.0` to all new functions to clarify their creation date.

Other than that, minor cleanups in places. I think there are some unrelated(?) Taxonomy changes that went in there.

**Comment by herbdool** [comment]
@docwilmot scheduled state is now saved. That part looks good now.

**Comment by herbdool** [comment]
Tested after most recent changes. Looks good to me.

**Comment by docwilmot** [comment]
@quicksketch all issues addressed in new commits.

To address your main concern, I propose a new `hook_layout_get_layout_by_path_alter(&$layouts, $router_item, $object)`, invoked in layout module's `layout_get_layout_by_path()` function, that allows other modules to modify the layouts that are loaded for a given path before the chosen layout is resolved. The `$object` parameter here would be the object provided to layouts based on contexts.

This is then implemented by Node to check if the router item path is the preview path and return the appropriate layout and node being previewed. 

I think this might be a bit better than special-casing preview paths directly in Layout module.

**Comment by quicksketch** [comment]
> Theres still this persistently failing date test that doesnt happen locally. 

Maybe try disabling Date module locally? The logic for handling date fields changes in the node form depending on whether it is enabled.

>  I propose a new hook_layout_get_layout_by_path_alter(&$layouts, $router_item, $object), invoked in layout module's layout_get_layout_by_path() function, that allows other modules to modify the layouts that are loaded for a given path before the chosen layout is resolved. 

Let's continue talking about options for this in https://github.com/backdrop/backdrop-issues/issues/1528, which is a closely related issue.

**Comment by quicksketch** [comment]
It looks like @docwilmot already created the new `hook_layout_get_layout_by_path_alter()` hook in the PR. I'd prefer we look at that problem separately in https://github.com/backdrop/backdrop-issues/issues/1528. So for now I think leaving the hack with a `@todo` in place is better than rushing into an API addition that might cause us trouble.

And for whatever reason, that date failure seems to be caused by these changes. We've got lots of other PRs being updated that don't have these same failures.

**Comment by docwilmot** [comment]
Pushed again
Still 4 failing Date tests. Still all green locally with or without date. I can not figure it out,

**Comment by quicksketch** [comment]
I'll see if I can suss out the source of the failure. We can't commit this if it's breaking the tests.

**Comment by herbdool** [comment]
I see the exceptions locally when I test it too.

**Comment by quicksketch** [comment]
Looks like the source of the problem is that our date fields don't handle default values properly when the fields are hidden. So for the sake of passing these tests, we just need to make the Created date field visible to the testing user by adding the "administer nodes" permission. We can make a follow-up for that new issue. I'll push a commit to the PR to fix up the tests.

**Comment by quicksketch** [comment]
I had some trouble rebasing the PR and somehow broke it... now it can't be reopened. I made a replacement PR at https://github.com/backdrop/backdrop/pull/2290.

**Comment by herbdool** [comment]
Tested again and it looks good.

**Comment by quicksketch** [comment]
Great. As this has already been RTBC 3 times, with these issues addressed we're good to go. I think we will need to tidy-up the CSS further, but we already knew that.

I've merged https://github.com/backdrop/backdrop/pull/2290 into 1.x for 1.11.0. Thanks everyone!
